macros/src/codegen.rs

```diff
@@ -224,106 +252,106 @@ pub fn generate_code(sm: &ParsedStateMachine) -> proc_macro2::TokenStream {
-
-            // get output state lifetimes
-            let out_state_lifetimes = sm.state_data.lifetimes.get(&value.out_state.to_string()).cloned().unwrap_or_default();
-
-            // get event lifetimes
-            let event_lifetimes = sm.event_data.lifetimes.get(event).cloned().unwrap_or_default();
-
-            // combine all lifetimes
-            let mut all_lifetimes = Lifetimes::new();
-            all_lifetimes.extend(&in_state_lifetimes);
-            all_lifetimes.extend(&out_state_lifetimes);
-            all_lifetimes.extend(&event_lifetimes);
-
-            // Create the guard traits for user implementation
-            if let Some(AsyncIdent {ident: guard, is_async}) = &value.guard {
-                let event_data = match sm.event_data.data_types.get(event) {
-                    Some(et @ Type::Reference(_)) => quote! { event_data: #et },
-                    Some(et) => quote! { event_data: &#et },
-                    None => quote! {},
-                };
-
-                let guard_error = if sm.custom_guard_error {
-                    quote! { Self::GuardError }
-                } else {
-                    quote! { () }
-                };
-
-                // Only add the guard if it hasn't been added before
-                if !guard_set.iter().any(|g| g == guard) {
-                    guard_set.push(guard.clone());
-                    let is_async = match is_async {
-                        true => quote!{ async },
-                        false => quote!{ },
-                    };
-                    guard_list.extend(quote! {
-                        #[allow(missing_docs)]
-                        #is_async fn #guard <#all_lifetimes> (&mut self, #temporary_context #state_data #event_data) -> Result<(), #guard_error>;
+        for (event, event_mapping) in event_mappings {
+            for transition in &event_mapping.transitions{
+                // get input state lifetimes
+                let in_state_lifetimes = sm.state_data.lifetimes.get(&event_mapping.in_state.to_string()).cloned().unwrap_or_default();
+
+                // get output state lifetimes
+                let out_state_lifetimes = sm.state_data.lifetimes.get(&transition.out_state.to_string()).cloned().unwrap_or_default();
+
+                // get event lifetimes
+                let event_lifetimes = sm.event_data.lifetimes.get(event).cloned().unwrap_or_default();
+
+                // combine all lifetimes
+                let mut all_lifetimes = Lifetimes::new();
+                all_lifetimes.extend(&in_state_lifetimes);
+                all_lifetimes.extend(&out_state_lifetimes);
+                all_lifetimes.extend(&event_lifetimes);
+
+                // Create the guard traits for user implementation
+                if let Some(guard_expression) = &transition.guard {
+                    let _ = visit_guards(guard_expression,|guard| {
```

Comment:
If we don't currently expect this to ever return an error, have it do an `unwrap()` instead so that we can catch faults if we accidentally add an Error code into this closure in the future.

Reply:
makes sense

Reply:
good idea, thanks, done